File: algorithms/36.swift
// https://leetcode.com/problems/valid-sudoku

class Solution {
    func isValidSudoku(board: [[Character]]) -> Bool {
        var M = [Int:[[Int]]]()
        for i in 0..<board.count {
            for j in 0..<board[i].count {
                let c = String(board[i][j])
                if c == "." { continue }

                let d = Int(c)!
                let g = i/3 * 3 + j/3
                if M.keys.contains(d) {
                    var rows = M[d]![0]
                    var cols = M[d]![1]
                    var groups = M[d]![2]
                    if rows.contains(i) || cols.contains(j) || groups.contains(g) { return false }
                    rows.append(i)
                    cols.append(j)
                    groups.append(g)
                    M[d] = [rows, cols, groups]
                } else {
                    M[d] = [[i], [j], [g]]
                }
            }
        }
        return true
    }
}

File: algorithms/36.py
# https://leetcode.com/problems/valid-sudoku


class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        d = {}
        for i in range(len(board)):
            for j in range(len(board[i])):
                c = board[i][j]
                if c == '.':
                    continue

                g = math.floor(i / 3) * 3 + math.floor(j / 3)
                if c in d:
                    rows, cols, groups = d[c][0], d[c][1], d[c][2]
                    if i in rows or j in cols or g in groups:
                        return False
                    rows.append(i)
                    cols.append(j)
                    groups.append(g)
                else:
                    d[c] = [[i], [j], [g]]
        return True
